src/test/integration/org/apache/cassandra/sidecar/routes/TokenRangeIntegrationLeavingTest.java

```diff
@@ -0,0 +1,786 @@
+    void retrieveMappingMultiDCHalveClusterSize(VertxTestContext context,
+                                                ConfigurableCassandraTestContext cassandraTestContext) throws Exception
+    {
+
+        int leavingNodesPerDC = 3;
+        CassandraIntegrationTest annotation = sidecarTestContext.cassandraTestContext().annotation;
+        int numNodes = annotation.nodesPerDc() + annotation.newNodesPerDc();
+        UpgradeableCluster cluster = getMultiDCCluster(numNodes,
+                                                       annotation.numDcs(),
+                                                       BBHelperHalveClusterMultiDC::install,
+                                                       cassandraTestContext);
+
+        Map<String, Map<Range<BigInteger>, List<String>>> expectedRangeMappings
+        = generateExpectedRangeHalveClusterSizeMultiDC();
+        runLeavingTestScenario(context,
+                               leavingNodesPerDC,
+                               BBHelperHalveClusterMultiDC.TRANSIENT_STATE_START,
+                               BBHelperHalveClusterMultiDC.TRANSIENT_STATE_END,
+                               cluster,
+                               expectedRangeMappings);
+    }
+
+    void runLeavingTestScenario(VertxTestContext context,
+                                ConfigurableCassandraTestContext cassandraTestContext,
+                                int leavingNodesPerDC,
+                                BiConsumer<ClassLoader, Integer> instanceInitializer,
+                                CountDownLatch transientStateStart,
+                                CountDownLatch transientStateEnd,
+                                Map<String, Map<Range<BigInteger>, List<String>>> expectedRangeMappings)
+    throws Exception
+    {
+
+        UpgradeableCluster cluster = cassandraTestContext.configureAndStartCluster(
+        builder -> builder.withInstanceInitializer(instanceInitializer));
+
+        runLeavingTestScenario(context,
+                               leavingNodesPerDC,
+                               transientStateStart,
+                               transientStateEnd,
+                               cluster,
+                               expectedRangeMappings);
+    }
+
+    void runLeavingTestScenario(VertxTestContext context,
+                                int leavingNodesPerDC,
+                                CountDownLatch transientStateStart,
+                                CountDownLatch transientStateEnd,
+                                UpgradeableCluster cluster,
+                                Map<String, Map<Range<BigInteger>, List<String>>> expectedRangeMappings)
+    throws Exception
+    {
+        try
+        {
+            CassandraIntegrationTest annotation = sidecarTestContext.cassandraTestContext().annotation;
+            Set<String> dcReplication;
+
+            if (annotation.numDcs() > 1)
```

Comment:
This logic is repeated a bunch of times - maybe move to the integration test base?